feat: one line command to convert YAMLs to HTML

dandhlee opened the pull request:
This features will allow a folder of just YAMLs to be quickly converted to HTMLs without interacting with Cloud Storage buckets, and will be done entirely on your local machine. Good for quickly testing to see if the YAML generates correct HTML.

To use this method, create a directory in doc-pipeline containing the YAMLs, like how they would show up before they are tar'd up and sent to Cloud Storage buckets. If we call that directory "docs_test" then from doc-pipeline, run
```INPUT=docs_test TRAMPOLINE_BUILD_FILE=./generate.sh TRAMPOLINE_IMAGE=gcr.io/cloud-devrel-kokoro-resources/docfx TRAMPOLINE_DOCKERFILE=docfx/Dockerfile ci/trampoline_v2.sh```
which would take the contents of `docs_test` directory, build the HTML pages, and return the directory of HTMLs. 

Visually, it would look something like this for the above example:

```
/doc-pipeline
...
|-- docs_test
|    |-- docs.metadata
|    |-- *.yml
```
builds to

```
/doc-pipeline
|-- docs_test
|    |-- docs.metadata
|    |-- *.yml
|    |-- pubsub
|    |    |-- *.html
|    |    |-- _toc.yaml
```
where pubsub would be the metadata name. 

It currently also works with `docs.metadata.json` and even if `docs_test` were to contain directories (api, example) and not have YAML files laid out flat.

Fixes #53

dandhlee committed:
feat: one command to convert YAML to HTML

dandhlee committed:
flake8 update

dandhlee commented:
Will be updating the README to reflect this change once approvals are given.

How does everyone feel about this feature? Future improvements include:
- caching pip install content so running this repeatedly isn't bottlenecked too much by all the package installations
- accepting input path for any local directories rather than having to move it into a directory within doc-pipeline

let me know!

jskeet commented:
I haven't looked at the code, but I'm very supportive of the general idea. I'd suggest not making it as yaml-focused in naming though. I'm assuming it would also generate HTML for any Markdown files we have (guides etc) - so possibly just `docs_test` or similar?

dandhlee commented:
It does process Markdown files and other structures that the normal pipeline supports. The name was irrelevant to the actual functionality of this feature, but to avoid the confusion I'll have that edited! The directory can be named anything you want, doesn't have to be `yaml_test`

jskeet commented:
Having had a look, I *think* local_generate.py is a modified copy of generate.py - is that right? If so, that worries me a bit in terms of staleness.

I'd prefer to have it as a sort of pipeline of common blocks, where the production pipeline involves downloading the tarball and uploading the tarball, but the local generation pipeline skips that and uses the given path... but using the same generation step. Is that feasible?

We'll need to work out how the xref stuff fits in with all of this as well...

dandhlee commented:
Incorporating this into the existing `generate.py` I thought might be adding too much noise to it for simple local generation, since most of it involves working with Cloud Storage buckets and whatnot, so I thought it'd be better to maintain this separately for local development. Perhaps the easier way would be to refactor further to minimize the strain adding this feature might cause to the existing pipeline.

Currently I've left out the part where it will add all the existing xrefs from CS bucket and simply work with xrefs given in the `docs.metadata*` file.

tbpg reviewed (commented):
+1 to @jskeet. This might take a bit of a refactor, depending on how things are set up. But, we should use the exact same code for generating locally as we do normally.

dandhlee commented:
Sounds good. Giving it some thought, I do agree that another file might be a bit too much to maintain, looking back at how much changes `generate.py` has gone through lately. I'll see what needs to happen to make this as clean change as possible.

Meanwhile, if there's any suggestions on how to improve the feature itself, please let me know!

tbpg commented:
Being in another file isn't necessarily the problem (and might even be a good thing) -- it's the duplication of logic.

dandhlee committed:
Merge branch 'master' of github.com:googleapis/doc-pipeline into easyhtml

dandhlee committed:
slight refactoring to reduce duplicate logic

dandhlee committed:
adding test case for local generation

dandhlee committed:
additional test coverage

dandhlee commented:
Refactored to de-duplicate the logic, left the local generation file on its own with a big of logic added to `generate.py` to help with local generation while reusing the original code as much as possible.

Added a test case as well for local generation.

Not sure how to handle xrefs for now. Any thoughts?

tbpg commented on generate.sh:
```diff
@@ -15,22 +15,26 @@
-elif [ -n "$SOURCE_BLOB" ]; then
-    python3 docpipeline/__main__.py build-one-doc $SOURCE_BUCKET $SOURCE_BLOB
+if [ -n "$INPUT" ]; then
+  python3 docpipeline/__main__.py build-local-doc $INPUT
```
Exit so you can drop the `else` right below this and keep indentation to a minimum.

```suggestion
  python3 docpipeline/__main__.py build-local-doc $INPUT
  exit 0
```

dandhlee replied:
done!

tbpg replied:
Looks like you need a push?

tbpg commented on docpipeline/generate.py:
```diff
@@ -80,6 +80,19 @@ def clone_templates(dir):
+        shutil.rmtree(templates_dir)
+    templates_dir.mkdir(parents=True, exist_ok=True)
+    log.info(f"Cloning templates into {templates_dir.absolute()}")
+    clone_templates(templates_dir)
```
Perhaps we should just inline `clone_templates`? Perhaps for a followup PR.

tbpg commented on docpipeline/generate.py:
```diff
@@ -137,9 +165,11 @@ def setup_docfx(tmp_path, blob):
-    metadata.xrefs[:] = [
-        get_xref(xref, blob.bucket, tmp_path) for xref in metadata.xrefs
-    ]
+    # TODO: Consider xrefs for local deployment
```
What happens if we try to `get_xref` when running locally? There is no bucket so it fails?

I think it would be better if we logged a warning about any xrefs that couldn't be fetched locally. In the future, we can implement a way for people to provide an `xrefs` directory where they've downloaded the xref files or something.

If we don't log a warning, the xrefs won't be processed and it could lead to confusing output.

Also, there are non-bucket xrefs we should probably still plumb through.

dandhlee replied:
any calls to a bucket class member would throw an exception when running locally, but taking a look at `get_xref` it simply would return an empty string for every xref so it shouldn't do much either way. If there are any xrefs stated in `docs.metadata*` it will retain the original list of xrefs supplied in the metadata. 

I think supplying a separate list of xrefs should be better if there are more xrefs to be considered.

tbpg replied:
Will you add a warning log message saying xrefs are being ignored?

dandhlee replied:
Done!

tbpg commented on docpipeline/generate.py:
```diff
@@ -108,26 +121,41 @@ def add_prettyprint(output_path):
             file_handle.write(html)
 
 
+def is_blob(blob):
```
I'm happy to move forward with this PR as-is. But, I wonder if there is a refactor we could do to avoid needing `is_blob`. Right now, `setup_docfx` does [something depending on blob/not] then common stuff. Then, `process_blob` does common stuff then [something depending on blob/not].

So, it seems like we could split those blob/not branches into separate functions.

dandhlee replied:
I've refactored it to local/bucket bits so that `is_blob` is no longer needed!

tbpg reviewed (changes requested):
A couple comments for this PR, a couple ideas for followup PRs. Looking good! Thank you!

dandhlee committed:
refactoring and making adjustments

dandhlee commented:
Updated! PTAL again :)

dandhlee committed:
Updating README

dandhlee commented:
@tbpg could you take one final look at the README? Thanks! 😄

tbpg commented on README.md:
```diff
@@ -114,7 +114,21 @@ black docpipeline tests
-### Running locally
+### Running locally for one package
+
+1. Create a directory on doc-pipeline (`my-dir`).
```
```suggestion
1. Create a directory inside `doc-pipeline` (for example, `my-dir`).
```

tbpg commented on README.md:
```diff
@@ -114,7 +114,21 @@ black docpipeline tests
+### Running locally for one package
+
+1. Create a directory on doc-pipeline (`my-dir`).
+1. Move the files necessary to build for one pacakage over to `my-dir`.
```
```suggestion
1. Create a `docs.metadata` file in `my-dir`. You can copy one from [here](https://github.com/googleapis/doc-pipeline/blob/master/testdata/docs.metadata).
1. Move or copy the `.yml` files for one package to `my-dir`.
```

tbpg commented on README.md:
```diff
@@ -114,7 +114,21 @@ black docpipeline tests
+1. The script runs `docfx build` over the package in my-dir, and places the
+   resulting HTML inside a subdirectory in `my-dir`. The subdirectory is
+   named after the package name found in the metadata.
+1. Note: running through this method will skip on processing xrefs other than
+   the ones found in the metadata.
```
I think it skips all xrefs?

dandhlee replied:
Confused between xrefs in `docs.metadata*` and xrefs that gets consumed in `docfx.json` by docfx. Updated.

dandhlee committed:
Update README.md

Co-authored-by: Tyler Bui-Palsulich <26876514+tbpg@users.noreply.github.com>

dandhlee committed:
Update README.md

Co-authored-by: Tyler Bui-Palsulich <26876514+tbpg@users.noreply.github.com>

dandhlee committed:
Updating README

dandhlee commented:
Done! PTAL again :)
